File: agents/python/src/agent/memory.py
"""Knowledge tools over the runbook library — the Ops Copilot's memory/RAG (Chapter 3.4).

The runbooks in ``agents/data/runbooks`` are the agent's long-term knowledge. ``get_runbook``
fetches one by its exact slug (an incident row carries its ``runbook`` slug); ``search_runbooks``
does a simple, deterministic keyword search when the slug is not known. Keyword retrieval keeps
the course fully offline — swap in a vector store or ADK ``MemoryService`` for semantic search.
"""

from __future__ import annotations

import logging
import re
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from google.adk.agents.llm_agent import ToolUnion

from . import data
from .config import settings
from .models import normalize_slug
from .resilience import with_resilience

logger = logging.getLogger(__name__)

# Terms shorter than this carry little retrieval signal, so drop them (a length filter, not a stop-list).
_MIN_TERM_LENGTH = 3


def _terms(query: str) -> list[str]:
    """Split a query into lowercase alphanumeric search terms, dropping very short ones."""
    return [term for term in re.findall(r"[a-z0-9]+", query.lower()) if len(term) >= _MIN_TERM_LENGTH]


def get_runbook(slug: str) -> dict[str, Any]:
    """Fetch a runbook by its exact slug (e.g. an incident's ``runbook`` field).

    Args:
        slug: The runbook identifier, e.g. ``high-latency`` or ``service-down``.

    Returns:
        A dict with the ``slug`` and its markdown ``content``, or an ``error`` if unknown.
    """
    # Parse at the boundary: normalize once here, then work only with the trusted
    # value — read, error message, and result all speak the normalized slug.
    normalized = normalize_slug(slug)
    if normalized is None:
        known = ", ".join(data.list_runbook_slugs())
        return {"error": f"Invalid runbook slug {slug!r}. Available runbooks: {known}."}
    content = data.read_runbook(normalized)
    if content is None:
        known = ", ".join(data.list_runbook_slugs())
        return {"error": f"No runbook named {normalized!r}. Available runbooks: {known}."}
    return {"slug": normalized, "content": content}


def search_runbooks(query: str, limit: int = 3) -> dict[str, Any]:
    """Search the runbook knowledge base for guidance relevant to a free-text query.

    Uses TF-IDF-style scoring by default: a term counts more when it is rare across
    runbooks (so ubiquitous words like "service" don't dominate), and a term matching
    a runbook's slug gets a strong boost. Deterministic and offline. With
    ``AGENT_SEMANTIC_RETRIEVAL=true`` it ranks by local-embedding similarity instead,
    falling back to this keyword scorer when the embedding model is unavailable.

    Args:
        query: What you are trying to resolve, e.g. ``database connection pool exhausted``.
        limit: The maximum number of runbooks to return (most relevant first).

    Returns:
        A dict with ``count`` and a ``runbooks`` list (slug + markdown content), best match first.
    """
    if limit <= 0:  # a non-positive limit falls back to the default
        limit = 3
    if settings.semantic_retrieval:
        # Imported lazily so the default offline path never touches the vector stack.
        from .retrieval import EmbeddingUnavailableError, semantic_search

        try:
            matches = semantic_search(query, limit)
            return {
                "count": len(matches),
                "retrieval": "semantic",
                "runbooks": [{"slug": row["slug"], "content": row["content"]} for row in matches],
            }
        except EmbeddingUnavailableError as error:
            # An explicit, logged fallback — never a silent degradation.
            logger.warning("Semantic retrieval unavailable, falling back to keywords: %s", error)
    terms = _terms(query)
    contents = {slug: (data.read_runbook(slug) or "") for slug in data.list_runbook_slugs()}
    total = len(contents) or 1
    # Document frequency: how many runbooks contain each distinct term.
    doc_freq = {term: sum(term in text.lower() for text in contents.values()) for term in set(terms)}

    scored: list[tuple[float, str, str]] = []
    for slug, content in contents.items():
        haystack = content.lower()
        score = 0.0
        for term in terms:
            frequency = doc_freq[term]
            if frequency:
                score += haystack.count(term) * (total / frequency)  # rarer term → higher weight
            if term in slug:
                score += total * 5  # the slug names the failure mode: a strong signal
        if score > 0:
            scored.append((score, slug, content))
    # Score descending, then slug ascending — the slug tie-break keeps ordering deterministic
    # for equal scores, which matters for reproducible evals.
    scored.sort(key=lambda row: (-row[0], row[1]))
    top = scored[:limit]
    return {"count": len(top), "runbooks": [{"slug": slug, "content": content} for _, slug, content in top]}


# The knowledge tools registered on the Ops Copilot (Ch. 3.4), wrapped with the
# same deadline/retry policy as the read tools — runbook reads are idempotent.
KNOWLEDGE_TOOLS: list[ToolUnion] = [with_resilience(get_runbook), with_resilience(search_runbooks)]

File: agents/go/internal/memory/memory.go
// Package memory provides knowledge tools over the runbook library — the Ops Copilot's
// memory/RAG (Chapter 3.4).
//
// The runbooks in agents/data/runbooks are the agent's long-term knowledge. getRunbook
// fetches one by its exact slug (an incident row carries its runbook slug); searchRunbooks
// does a simple, deterministic TF-IDF keyword search. Keyword retrieval keeps the course
// fully offline — swap in a vector store or ADK MemoryService for semantic search.
package memory

import (
	"fmt"
	"regexp"
	"sort"
	"strings"

	"google.golang.org/adk/v2/agent"
	"google.golang.org/adk/v2/tool"
	"google.golang.org/adk/v2/tool/functiontool"

	"github.com/MLOps-Courses/agentops-open-course/agents/go/internal/data"
)

// minTermLength drops very short, low-signal words from a query.
const minTermLength = 3

// defaultLimit is how many runbooks searchRunbooks returns when the caller passes zero.
const defaultLimit = 3

var termPattern = regexp.MustCompile(`[a-z0-9]+`)

// terms splits a query into lowercase search terms, dropping very short ones.
func terms(query string) []string {
	var out []string
	for _, term := range termPattern.FindAllString(strings.ToLower(query), -1) {
		if len(term) >= minTermLength {
			out = append(out, term)
		}
	}
	return out
}

type getRunbookInput struct {
	Slug string `json:"slug"` // runbook identifier, e.g. high-latency or service-down
}

type searchRunbooksInput struct {
	Query string `json:"query"` // what you are trying to resolve, e.g. "database pool exhausted"
	Limit int    `json:"limit"` // max runbooks to return (default 3 when zero)
}

type runbook struct {
	Slug    string `json:"slug"`
	Content string `json:"content"`
}

type searchRunbooksOutput struct {
	Count    int       `json:"count"`
	Runbooks []runbook `json:"runbooks"`
}

// KnowledgeTools builds the memory/RAG tools registered on the agent (Ch. 3.4).
func KnowledgeTools(store *data.Store) ([]tool.Tool, error) {
	all := make([]tool.Tool, 0, 2)
	for _, build := range []func(*data.Store) (tool.Tool, error){getRunbookTool, searchRunbooksTool} {
		tl, err := build(store)
		if err != nil {
			return nil, err
		}
		all = append(all, tl)
	}
	return all, nil
}

func getRunbookTool(store *data.Store) (tool.Tool, error) {
	return functiontool.New(functiontool.Config{
		Name:        "get_runbook",
		Description: "Fetch a runbook by its exact slug (e.g. an incident's runbook field).",
	}, func(_ agent.Context, in getRunbookInput) (map[string]any, error) {
		content, ok, err := store.ReadRunbook(in.Slug)
		if err != nil {
			return nil, err
		}
		if !ok {
			slugs, listErr := store.ListRunbookSlugs()
			if listErr != nil {
				return nil, listErr
			}
			return map[string]any{
				"error": fmt.Sprintf("No runbook named %q. Available runbooks: %s.", in.Slug, strings.Join(slugs, ", ")),
			}, nil
		}
		return map[string]any{"slug": in.Slug, "content": content}, nil
	})
}

func searchRunbooksTool(store *data.Store) (tool.Tool, error) {
	return functiontool.New(functiontool.Config{
		Name: "search_runbooks",
		Description: "Search the runbook knowledge base for guidance relevant to a free-text query " +
			"(TF-IDF keyword scoring, most relevant first).",
	}, func(_ agent.Context, in searchRunbooksInput) (searchRunbooksOutput, error) {
		return searchRunbooks(store, in.Query, in.Limit)
	})
}

// searchRunbooks ranks runbooks with TF-IDF-style scoring: rarer terms weigh more (so
// ubiquitous words like "service" don't dominate) and a term matching a runbook's slug
// gets a strong boost.
func searchRunbooks(store *data.Store, query string, limit int) (searchRunbooksOutput, error) {
	if limit <= 0 {
		limit = defaultLimit
	}
	slugs, err := store.ListRunbookSlugs()
	if err != nil {
		return searchRunbooksOutput{}, err
	}
	contents := make(map[string]string, len(slugs))
	for _, slug := range slugs {
		content, _, readErr := store.ReadRunbook(slug)
		if readErr != nil {
			return searchRunbooksOutput{}, readErr
		}
		contents[slug] = content
	}

	total := len(contents)
	if total == 0 {
		return searchRunbooksOutput{Count: 0, Runbooks: []runbook{}}, nil
	}
	queryTerms := terms(query)
	docFreq := documentFrequency(queryTerms, contents)

	type ranked struct {
		slug  string
		score float64
	}
	var results []ranked
	for slug, content := range contents {
		haystack := strings.ToLower(content)
		var score float64
		for _, term := range queryTerms {
			if freq := docFreq[term]; freq > 0 {
				score += float64(strings.Count(haystack, term)) * (float64(total) / float64(freq))
			}
			if strings.Contains(slug, term) {
				score += float64(total) * 5 // the slug names the failure mode: a strong signal
			}
		}
		if score > 0 {
			results = append(results, ranked{slug: slug, score: score})
		}
	}
	// Score descending, then slug ascending for deterministic ties.
	sort.Slice(results, func(i, j int) bool {
		if results[i].score != results[j].score {
			return results[i].score > results[j].score
		}
		return results[i].slug < results[j].slug
	})

	out := searchRunbooksOutput{Runbooks: []runbook{}}
	for i, r := range results {
		if i >= limit {
			break
		}
		out.Runbooks = append(out.Runbooks, runbook{Slug: r.slug, Content: contents[r.slug]})
	}
	out.Count = len(out.Runbooks)
	return out, nil
}

// documentFrequency counts, for each distinct term, how many runbooks contain it.
func documentFrequency(queryTerms []string, contents map[string]string) map[string]int {
	docFreq := map[string]int{}
	for _, term := range queryTerms {
		if _, seen := docFreq[term]; seen {
			continue
		}
		count := 0
		for _, content := range contents {
			if strings.Contains(strings.ToLower(content), term) {
				count++
			}
		}
		docFreq[term] = count
	}
	return docFreq
}
